File: framec_tests/javascript/src/match/match.js
// emitted from framec_v0.10.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

function FrameEvent(message, parameters) {

    var that = {};
    that._message = message;
    that._parameters = parameters;
    that._return = null;
    return that;
    
}

class Match {
    
    // creating private properties
    
    #state
    #compartment
    #nextCompartment
    
    constructor () {
        
        // Create and intialize start state compartment.
        
        this.#state = this.#sInit_;
        this.#compartment = new MatchCompartment(this.#state);
        this.#nextCompartment = null;
        
        // Initialize domain
        this.tape = [];
        
        // Send system start event
        const frameEvent = FrameEvent(">", null);
        this.#mux(frameEvent);
    }
    
    //===================== Interface Block ===================//
    
    Empty() {
        let e = FrameEvent("Empty",null);
        this.#mux(e);
    }
    
    Simple() {
        let e = FrameEvent("Simple",null);
        this.#mux(e);
    }
    
    Multi() {
        let e = FrameEvent("Multi",null);
        this.#mux(e);
    }
    
    Nested() {
        let e = FrameEvent("Nested",null);
        this.#mux(e);
    }
    
    Child() {
        let e = FrameEvent("Child",null);
        this.#mux(e);
    }
    
    OnInt(i) {
        let e = FrameEvent("OnInt",{"i":i});
        this.#mux(e);
    }
    
    OnString(s) {
        let e = FrameEvent("OnString",{"s":s});
        this.#mux(e);
    }
    
    //====================== Multiplexer ====================//
    
    #mux(e) {
        switch (this.#compartment.state) {
            case this.#sInit_:
                this.#sInit_(e);
                break;
            case this.#sEmptyMatch_:
                this.#sEmptyMatch_(e);
                break;
            case this.#sSimpleMatch_:
                this.#sSimpleMatch_(e);
                break;
            case this.#sMultiMatch_:
                this.#sMultiMatch_(e);
                break;
            case this.#sNestedMatch_:
                this.#sNestedMatch_(e);
                break;
            case this.#sChildMatch_:
                this.#sChildMatch_(e);
                break;
            case this.#sFinal_:
                this.#sFinal_(e);
                break;
        }
        
        if( this.#nextCompartment != null) {
            let nextCompartment = this.#nextCompartment
            this.#nextCompartment = null
            if (nextCompartment._forwardEvent != null && 
               nextCompartment._forwardEvent._message == ">") {
                this.#mux(FrameEvent( "<", this.#compartment.ExitArgs))
                this.#compartment = nextCompartment
                this.#mux(nextCompartment._forwardEvent)
            } else {
                this.#doTransition(nextCompartment)
                if (nextCompartment._forwardEvent != null) {
                    this.#mux(nextCompartment._forwardEvent)
                }
            }
            nextCompartment._forwardEvent = null
        }
    }
    
    //===================== Machine Block ===================//
    
    #sInit_(e) {
        switch (e._message) {
            case "Empty":
                {
                let compartment =  new MatchCompartment(this.#sEmptyMatch_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
            case "Simple":
                {
                let compartment =  new MatchCompartment(this.#sSimpleMatch_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
            case "Multi":
                {
                let compartment =  new MatchCompartment(this.#sMultiMatch_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
            case "Nested":
                {
                let compartment =  new MatchCompartment(this.#sNestedMatch_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
            case "Child":
                {
                let compartment =  new MatchCompartment(this.#sChildMatch_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
        }
    }
    
    #sEmptyMatch_(e) {
        switch (e._message) {
            case "OnString":
                {
                if (((e._parameters["s"]) == "") || ((e._parameters["s"]) == "foo")) {
                    this.log_do("empty");
                } else {
                    this.log_do("?");
                }
                
                return;
                }
                
        }
    }  //  TODO: matching only the empty string is broken
	
    
    #sSimpleMatch_(e) {
        switch (e._message) {
            case "OnInt":
                {
                if (((e._parameters["i"]) == 0)) {
                    this.log_do("0");
                } else if (((e._parameters["i"]) == 42)) {
                    this.log_do("42");
                } else if (((e._parameters["i"]) == 42)) {
                    this.log_do("!!!");
                } else if (((e._parameters["i"]) == -200)) {
                    this.log_do("-200");
                } else {
                    this.log_do("?");
                }
                
                return;
                }
                
            case "OnString":
                {
                if (((e._parameters["s"]) == "hello")) {
                    this.log_do("hello");
                } else if (((e._parameters["s"]) == "hello")) {
                    this.log_do("!!!");
                } else if (((e._parameters["s"]) == "goodbye")) {
                    this.log_do("goodbye");
                } else if (((e._parameters["s"]) == "Testing 1, 2, 3...")) {
                    this.log_do("testing");
                } else if (((e._parameters["s"]) == "$10!")) {
                    this.log_do("money");
                } else {
                    this.log_do("?");
                }
                
                return;
                }
                
        }
    }
    
    #sMultiMatch_(e) {
        switch (e._message) {
            case "OnInt":
                {
                if (((e._parameters["i"]) == 3) || ((e._parameters["i"]) == -7)) {
                    this.log_do("3|-7");
                } else if (((e._parameters["i"]) == -4) || ((e._parameters["i"]) == 5) || ((e._parameters["i"]) == 6)) {
                    this.log_do("-4|5|6");
                } else {
                    this.log_do("?");
                }
                
                return;
                }
                
            case "OnString":
                {
                if (((e._parameters["s"]) == "$10") || ((e._parameters["s"]) == "12.5%") || ((e._parameters["s"]) == "@#*!")) {
                    this.log_do("symbols");
                } else if (((e._parameters["s"]) == " ") || ((e._parameters["s"]) == "  ") || ((e._parameters["s"]) == "\t") || ((e._parameters["s"]) == "\n")) {
                    this.log_do("whitespace");
                } else {
                    this.log_do("?");
                }
                
                return;
                }
                
        }
    }
    
    #sNestedMatch_(e) {
        switch (e._message) {
            case "OnInt":
                {
                if ((e._parameters["i"]) > 0) {
                    if (((e._parameters["i"]) == 1) || ((e._parameters["i"]) == 2) || ((e._parameters["i"]) == 3)) {
                        this.log_do("1-3");
                        if (((e._parameters["i"]) == 1)) {
                            this.log_do("1");
                        } else if (((e._parameters["i"]) == 2)) {
                            this.log_do("2");
                        } else {
                            this.log_do("3");
                        }
                    } else if (((e._parameters["i"]) == 4) || ((e._parameters["i"]) == 5)) {
                        this.log_do("4-5");
                        if ((e._parameters["i"]) == 4) {
                            this.log_do("4");
                        } else {
                            this.log_do("5");
                        }
                    } else {
                        this.log_do("too big");
                    }
                } else {
                    this.log_do("too small");
                }
                
                return;
                }
                
            case "OnString":
                {
                if (((e._parameters["s"]) == "hello") || ((e._parameters["s"]) == "hola") || ((e._parameters["s"]) == "bonjour")) {
                    this.log_do("greeting");
                    if (((e._parameters["s"]) == "hello")) {
                        this.log_do("English");
                    } else if (((e._parameters["s"]) == "hola")) {
                        this.log_do("Spanish");
                    } else {
                        this.log_do("French");
                    }
                } else if (((e._parameters["s"]) == "goodbye") || ((e._parameters["s"]) == "adios") || ((e._parameters["s"]) == "au revoir")) {
                    this.log_do("farewell");
                    if (((e._parameters["s"]) == "goodbye")) {
                        this.log_do("English");
                    } else if (((e._parameters["s"]) == "adios")) {
                        this.log_do("Spanish");
                    } else {
                        this.log_do("French");
                    }
                } else {
                    this.log_do("?");
                }
                
                return;
                }
                
        }
    }
    
    #sChildMatch_(e) {
        switch (e._message) {
            case "OnInt":
                {
                if (((e._parameters["i"]) == 0)) {
                    let compartment =  new MatchCompartment(this.#sFinal_);
                    
                    
                    this.#transition(compartment);
                    return;
                } else if (((e._parameters["i"]) == 3)) {
                    this.log_do("3");
                } else if (((e._parameters["i"]) == 4)) {
                    this.log_do("4");
                    
                    return;
                } else if (((e._parameters["i"]) == 42)) {
                    this.log_do("42 in child");
                } else if (((e._parameters["i"]) == 5)) {
                    this.log_do("5");
                    let compartment =  new MatchCompartment(this.#sFinal_);
                    
                    
                    this.#transition(compartment);
                    return;
                } else {
                    this.log_do("no match in child");
                }
                
                break;
                }
                
            case "OnString":
                {
                if (((e._parameters["s"]) == "hello")) {
                    this.log_do("hello in child");
                } else if (((e._parameters["s"]) == "goodbye")) {
                    let compartment =  new MatchCompartment(this.#sFinal_);
                    
                    
                    this.#transition(compartment);
                    return;
                } else if (((e._parameters["s"]) == "Testing 1, 2, 3...")) {
                    this.log_do("testing in child");
                    
                    return;
                } else {
                    this.log_do("no match in child");
                }
                
                break;
                }
                
        }
        this.#sSimpleMatch_(e);
        
    }
    
    #sFinal_(e) {
        switch (e._message) {
        }
    }
    
    //===================== Actions Block ===================//
    
    // Unimplemented Actions
    
    log_do(msg) { throw new Error('Action not implemented.'); }
    
    //=============== Machinery and Mechanisms ==============//
    
    #transition(compartment) {
        this.#nextCompartment = compartment;
    }
    
    #doTransition(nextCompartment) {
        this.#mux(FrameEvent("<", this.#compartment.ExitArgs));
        this.#compartment = nextCompartment;
        this.#mux(FrameEvent(">", this.#compartment.EnterArgs));
    }
    
    state_info() {
        return this.#compartment.state.name;
    }
    
    
};

//=============== Compartment ==============//

class MatchCompartment {

    constructor(state) {
        this.state = state
    }
    
    StateArgs = {};
    StateVars = {};
    EnterArgs = {};
    ExitArgs = {};
    _forwardEvent = FrameEvent.call(this)
}


/********************

class MatchController extends Match {

	constructor() {
	  super()
	}
	log_do(msg) {}
};

********************/

module.exports = Match

File: framec_tests/java/Match/Match.java
// emitted from framec_v0.10.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

package framec_tests.java.Match;
import java.util.*;
import framec_tests.java.FrameLang.FrameEvent;

class Match {
    
    private MatchCompartment _compartment_;
    private MatchCompartment _nextCompartment_;
    
    
    
    Match() {
        
        // Create and intialize start state compartment.
        this._compartment_ = new MatchCompartment(this._state_);
        this._nextCompartment_ = null;
        
        
        // Send system start event
        FrameEvent frame_event = new FrameEvent(">", null);
        this._mux_(frame_event);
    
    }
    
    // states enum
    private enum MatchState {
        
        INIT(0),
        EMPTYMATCH(1),
        SIMPLEMATCH(2),
        MULTIMATCH(3),
        NESTEDMATCH(4),
        CHILDMATCH(5),
        FINAL(6);
        
        public final int value;
        
        private MatchState(int value) {
            this.value=value;
        }
        
        public int getValue() {
            return value;
        }
    }
    
    //====================== Multiplexer ====================//
    
    private void _mux_(FrameEvent e) {
        if(this._compartment_.state == MatchState.INIT.getValue()) {
            this._sInit_(e);
        }else if(this._compartment_.state == MatchState.EMPTYMATCH.getValue()) {
            this._sEmptyMatch_(e);
        }else if(this._compartment_.state == MatchState.SIMPLEMATCH.getValue()) {
            this._sSimpleMatch_(e);
        }else if(this._compartment_.state == MatchState.MULTIMATCH.getValue()) {
            this._sMultiMatch_(e);
        }else if(this._compartment_.state == MatchState.NESTEDMATCH.getValue()) {
            this._sNestedMatch_(e);
        }else if(this._compartment_.state == MatchState.CHILDMATCH.getValue()) {
            this._sChildMatch_(e);
        }else if(this._compartment_.state == MatchState.FINAL.getValue()) {
            this._sFinal_(e);
        }
        
        
        if(this._nextCompartment_ != null) {
            MatchCompartment nextCompartment = this._nextCompartment_;
            this._nextCompartment_ = null;
            if(nextCompartment._forwardEvent != null && 
             nextCompartment._forwardEvent._message == ">") {
                this._mux_(new FrameEvent( "<", this._compartment_.exitArgs));
                this._compartment_ = nextCompartment;
                this._mux_(nextCompartment._forwardEvent);
            } else {
                this._doTransition_(nextCompartment);
                if(nextCompartment._forwardEvent != null) {
                    this._mux_(nextCompartment._forwardEvent);
                }
            }
            nextCompartment._forwardEvent = null;
        }
    }
    
    //===================== Interface Block ===================//
    
    public void Empty() {
        FrameEvent e = new FrameEvent("Empty", null);
        this._mux_(e);
    }
    
    public void Simple() {
        FrameEvent e = new FrameEvent("Simple", null);
        this._mux_(e);
    }
    
    public void Multi() {
        FrameEvent e = new FrameEvent("Multi", null);
        this._mux_(e);
    }
    
    public void Nested() {
        FrameEvent e = new FrameEvent("Nested", null);
        this._mux_(e);
    }
    
    public void Child() {
        FrameEvent e = new FrameEvent("Child", null);
        this._mux_(e);
    }
    
    public void OnInt(int i) {
        HashMap<String,Object> parameters = new HashMap<String,Object>();
        parameters.put("i", i);

        FrameEvent e = new FrameEvent("OnInt", parameters);
        this._mux_(e);
    }
    
    public void OnString(String s) {
        HashMap<String,Object> parameters = new HashMap<String,Object>();
        parameters.put("s", s);

        FrameEvent e = new FrameEvent("OnString", parameters);
        this._mux_(e);
    }
    
    
    //===================== Machine Block ===================//
    
    private void _sInit_(FrameEvent e) {
        if(e._message == "Empty") {
            MatchCompartment compartment =  new MatchCompartment(MatchState.EMPTYMATCH.getValue());
            
            this._transition_(compartment);
            
            return;
        }
        else if(e._message == "Simple") {
            MatchCompartment compartment =  new MatchCompartment(MatchState.SIMPLEMATCH.getValue());
            
            this._transition_(compartment);
            
            return;
        }
        else if(e._message == "Multi") {
            MatchCompartment compartment =  new MatchCompartment(MatchState.MULTIMATCH.getValue());
            
            this._transition_(compartment);
            
            return;
        }
        else if(e._message == "Nested") {
            MatchCompartment compartment =  new MatchCompartment(MatchState.NESTEDMATCH.getValue());
            
            this._transition_(compartment);
            
            return;
        }
        else if(e._message == "Child") {
            MatchCompartment compartment =  new MatchCompartment(MatchState.CHILDMATCH.getValue());
            
            this._transition_(compartment);
            
            return;
        }
    }
    
    private void _sEmptyMatch_(FrameEvent e) {
        if(e._message == "OnString") {
            if ((((String) e._parameters.get("s")) == "") || (((String) e._parameters.get("s")) == "foo")) {
                log_do("empty");
            } else {
                log_do("?");
            }
            
            return;
        }
    }  //  TODO: matching only the empty string is broken
    
    
    private void _sSimpleMatch_(FrameEvent e) {
        if(e._message == "OnInt") {
            if ((((int) e._parameters.get("i")) == 0)) {
                log_do("0");
            } else if ((((int) e._parameters.get("i")) == 42)) {
                log_do("42");
            } else if ((((int) e._parameters.get("i")) == 42)) {
                log_do("!!!");
            } else if ((((int) e._parameters.get("i")) == -200)) {
                log_do("-200");
            } else {
                log_do("?");
            }
            
            return;
        }
        else if(e._message == "OnString") {
            if ((((String) e._parameters.get("s")) == "hello")) {
                log_do("hello");
            } else if ((((String) e._parameters.get("s")) == "hello")) {
                log_do("!!!");
            } else if ((((String) e._parameters.get("s")) == "goodbye")) {
                log_do("goodbye");
            } else if ((((String) e._parameters.get("s")) == "Testing 1, 2, 3...")) {
                log_do("testing");
            } else if ((((String) e._parameters.get("s")) == "$10!")) {
                log_do("money");
            } else {
                log_do("?");
            }
            
            return;
        }
    }
    
    private void _sMultiMatch_(FrameEvent e) {
        if(e._message == "OnInt") {
            if ((((int) e._parameters.get("i")) == 3) || (((int) e._parameters.get("i")) == -7)) {
                log_do("3|-7");
            } else if ((((int) e._parameters.get("i")) == -4) || (((int) e._parameters.get("i")) == 5) || (((int) e._parameters.get("i")) == 6)) {
                log_do("-4|5|6");
            } else {
                log_do("?");
            }
            
            return;
        }
        else if(e._message == "OnString") {
            if ((((String) e._parameters.get("s")) == "$10") || (((String) e._parameters.get("s")) == "12.5%") || (((String) e._parameters.get("s")) == "@#*!")) {
                log_do("symbols");
            } else if ((((String) e._parameters.get("s")) == " ") || (((String) e._parameters.get("s")) == "  ") || (((String) e._parameters.get("s")) == "\t") || (((String) e._parameters.get("s")) == "\n")) {
                log_do("whitespace");
            } else {
                log_do("?");
            }
            
            return;
        }
    }
    
    private void _sNestedMatch_(FrameEvent e) {
        if(e._message == "OnInt") {
            if (((int) e._parameters.get("i")) > 0) {
                if ((((int) e._parameters.get("i")) == 1) || (((int) e._parameters.get("i")) == 2) || (((int) e._parameters.get("i")) == 3)) {
                    log_do("1-3");
                    if ((((int) e._parameters.get("i")) == 1)) {
                        log_do("1");
                    } else if ((((int) e._parameters.get("i")) == 2)) {
                        log_do("2");
                    } else {
                        log_do("3");
                    }
                } else if ((((int) e._parameters.get("i")) == 4) || (((int) e._parameters.get("i")) == 5)) {
                    log_do("4-5");
                    if (((int) e._parameters.get("i")) == 4) {
                        log_do("4");
                    } else {
                        log_do("5");
                    }
                } else {
                    log_do("too big");
                }
            } else {
                log_do("too small");
            }
            
            return;
        }
        else if(e._message == "OnString") {
            if ((((String) e._parameters.get("s")) == "hello") || (((String) e._parameters.get("s")) == "hola") || (((String) e._parameters.get("s")) == "bonjour")) {
                log_do("greeting");
                if ((((String) e._parameters.get("s")) == "hello")) {
                    log_do("English");
                } else if ((((String) e._parameters.get("s")) == "hola")) {
                    log_do("Spanish");
                } else {
                    log_do("French");
                }
            } else if ((((String) e._parameters.get("s")) == "goodbye") || (((String) e._parameters.get("s")) == "adios") || (((String) e._parameters.get("s")) == "au revoir")) {
                log_do("farewell");
                if ((((String) e._parameters.get("s")) == "goodbye")) {
                    log_do("English");
                } else if ((((String) e._parameters.get("s")) == "adios")) {
                    log_do("Spanish");
                } else {
                    log_do("French");
                }
            } else {
                log_do("?");
            }
            
            return;
        }
    }
    
    private void _sChildMatch_(FrameEvent e) {
        if(e._message == "OnInt") {
            if ((((int) e._parameters.get("i")) == 0)) {
                MatchCompartment compartment =  new MatchCompartment(MatchState.FINAL.getValue());
                
                this._transition_(compartment);
            } else if ((((int) e._parameters.get("i")) == 3)) {
                log_do("3");
            } else if ((((int) e._parameters.get("i")) == 4)) {
                log_do("4");
                
                return;
            } else if ((((int) e._parameters.get("i")) == 42)) {
                log_do("42 in child");
            } else if ((((int) e._parameters.get("i")) == 5)) {
                log_do("5");
                MatchCompartment compartment =  new MatchCompartment(MatchState.FINAL.getValue());
                
                this._transition_(compartment);
                
                return;
            } else {
                log_do("no match in child");
            }
            
        }
        else if(e._message == "OnString") {
            if ((((String) e._parameters.get("s")) == "hello")) {
                log_do("hello in child");
            } else if ((((String) e._parameters.get("s")) == "goodbye")) {
                MatchCompartment compartment =  new MatchCompartment(MatchState.FINAL.getValue());
                
                this._transition_(compartment);
            } else if ((((String) e._parameters.get("s")) == "Testing 1, 2, 3...")) {
                log_do("testing in child");
                
                return;
            } else {
                log_do("no match in child");
            }
            
        }
        _sSimpleMatch_(e);
        
    }
    
    private void _sFinal_(FrameEvent e) {
    }
    
    //===================== Actions Block ===================//
    
    // Unimplemented Actions
    
    protected void log_do(String msg) { throw new UnsupportedOperationException(); }
    
    //===================== Domain Block ===================//
    
    public ArrayList<String> tape  = new ArrayList<String>();
    
    //=============== Machinery and Mechanisms ==============//
    
    private int _state_;
    
    private void _transition_(MatchCompartment compartment) {
        this._nextCompartment_ = compartment;
    }
    
    private void _doTransition_(MatchCompartment nextCompartment) {
        this._mux_(new FrameEvent("<", this._compartment_.exitArgs));
        this._compartment_ = nextCompartment;
        this._mux_(new FrameEvent(">", this._compartment_.enterArgs));
    }
    
    public String state_info(){
        return String.valueOf(this._compartment_.state);
        }
        
}

//=============== Compartment ==============//

class MatchCompartment {

    int state;
    
    MatchCompartment(int state) {
        this.state = state;
    }
    
    HashMap<String, Object> stateArgs = new HashMap<String, Object>();
    HashMap<String, Object> stateVars = new HashMap<String, Object>();
    HashMap<String, Object> enterArgs = new HashMap<String, Object>();
    HashMap<String, Object> exitArgs = new HashMap<String, Object>();
    FrameEvent _forwardEvent = new FrameEvent();
}


/********************

class MatchController extends Match {

	MatchController() {
	  super();
	}

    protected void log_do(String msg) {}
}

********************/

